File: src/main/java/com/lcaohoanq/nocket/domain/reaction/ReactionController.java
package com.lcaohoanq.nocket.domain.reaction;

import com.lcaohoanq.nocket.api.ApiResponse;
import com.lcaohoanq.nocket.base.exception.DataNotFoundException;
import com.lcaohoanq.nocket.enums.EReaction;
import com.lcaohoanq.nocket.mapper.ReactionMapper;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("${api.prefix}/reactions")
@RequiredArgsConstructor
public class ReactionController {

    private final ReactionRepository reactionRepository;
    private final ReactionMapper reactionMapper;

    @GetMapping("/all")
    public ResponseEntity<ApiResponse<List<ReactionPort.ReactionResponse>>> getAllReactions() {
        return ResponseEntity.ok(
            ApiResponse.<List<ReactionPort.ReactionResponse>>builder()
                .message("Get all reactions success")
                .statusCode(HttpStatus.OK.value())
                .isSuccess(true)
                .data(reactionRepository.findAll().stream().map(reactionMapper::toReactionResponse)
                          .toList())
                .build()
        );
    }

    @GetMapping("/detail")
    public ResponseEntity<ApiResponse<ReactionPort.ReactionResponse>> getReactionById(
        @RequestParam UUID id
    ) {
        return ResponseEntity.ok(
            ApiResponse.<ReactionPort.ReactionResponse>builder()
                .message("Get reaction by id success")
                .statusCode(HttpStatus.OK.value())
                .isSuccess(true)
                .data(reactionRepository.findById(id).map(reactionMapper::toReactionResponse)
                          .orElseThrow(
                              () -> new DataNotFoundException("Reaction not found")
                          ))
                .build()
        );
    }

    @PostMapping("/generate-reaction")
    @PreAuthorize("hasAnyRole('ROLE_MANAGER', 'ROLE_MEMBER', 'ROLE_STAFF')")
    public ResponseEntity<ApiResponse<ReactionPort.ReactionResponse>> generateReaction(
        @RequestBody ReactionPort.ReactionDTO reaction
    ) {

        Reaction newReaction = new Reaction();
        newReaction.setReaction(reaction.getReaction());

        return ResponseEntity.ok(
            ApiResponse.<ReactionPort.ReactionResponse>builder()
                .message("Generate reaction success")
                .statusCode(HttpStatus.OK.value())
                .isSuccess(true)
                .data(reactionMapper.toReactionResponse(reactionRepository.save(newReaction)))
                .build()
        );
    }
}

File: server/src/main/kotlin/com/lcaohoanq/nocket/domain/reaction/ReactionController.kt
package com.lcaohoanq.nocket.domain.reaction

import com.lcaohoanq.nocket.api.ApiResponse
import com.lcaohoanq.nocket.base.exception.DataNotFoundException
import com.lcaohoanq.nocket.domain.reaction.ReactionPort.ReactionDTO
import com.lcaohoanq.nocket.domain.reaction.ReactionPort.ReactionResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.util.*

@RequestMapping("\${api.prefix}/reactions")
@RestController
class ReactionController(
    private val reactionRepository: ReactionRepository,
    private val reactionMapper: ReactionMapper,
) {

    @GetMapping("/all")
    fun getAllReactions(): ResponseEntity<ApiResponse<List<ReactionResponse>>> = ResponseEntity.ok(
        ApiResponse.builder<List<ReactionResponse>>()
            .message("Get all reactions success")
            .statusCode(HttpStatus.OK.value())
            .isSuccess(true)
            .data(
                reactionRepository.findAll().stream()
                    .map { reaction: Reaction? -> reactionMapper.toReactionResponse(reaction) }
                    .toList())
            .build()
    )

    @GetMapping("/detail")
    fun getReactionById(
        @RequestParam id: UUID
    ): ResponseEntity<ApiResponse<ReactionResponse>> = ResponseEntity.ok(
        ApiResponse.builder<ReactionResponse>()
            .message("Get reaction by id success")
            .statusCode(HttpStatus.OK.value())
            .isSuccess(true)
            .data(
                reactionRepository.findById(id).map { reaction: Reaction? ->
                    reactionMapper.toReactionResponse(
                        reaction
                    )
                }
                    .orElseThrow { DataNotFoundException("Reaction not found") })
            .build()
    )

    @PostMapping("/generate-reaction")
    @PreAuthorize("hasAnyRole('ROLE_MANAGER', 'ROLE_MEMBER', 'ROLE_STAFF')")
    fun generateReaction(
        @RequestBody reaction: ReactionDTO
    ): ResponseEntity<ApiResponse<ReactionResponse>> {
        val newReaction = Reaction()
        newReaction.setReaction(reaction.reaction)

        return ResponseEntity.ok(
            ApiResponse.builder<ReactionResponse>()
                .message("Generate reaction success")
                .statusCode(HttpStatus.OK.value())
                .isSuccess(true)
                .data(reactionMapper.toReactionResponse(reactionRepository.save(newReaction)))
                .build()
        )
    }
}
